### PR TITLE
tracer: Fix race in spanContext.setSamplingPriority

### PR DESCRIPTION
**Pull request by evanj**
### What does this PR do?

This function read c.trace.priority without holding a lock, then called functions that modify it. This causes the race detector to detect the following race. This race bug is causing flaky tests in a program that uses dd-trace-go. The race is triggered when two goroutines both set the ext.ManualKeep on spans that share a root. I think it might plausibly cause a real bug bug, where c.updated gets set to true, even though it the priority was not changed.

My attempt to fix it is to move the logic to detect if the priority was modified into the setSamplingPriority function itself. This might not be the right fix since I don't really understand what this does, but it seems to pass the tests both with and without race detection.

The race detector report without this change:
```
WARNING: DATA RACE
Write at 0x00c0019a0a88 by goroutine 2435:
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*trace).setSamplingPriorityLocked()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/spancontext.go:327 +0xc4
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*trace).push()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/spancontext.go:359 +0x23c
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newSpanContext()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/spancontext.go:143 +0x904
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).StartSpan()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/tracer.go:516 +0xba0
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).newChildSpan()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/tracer_test.go:52 +0x94
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.TestTraceManualKeepRace.func1.1()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/span_test.go:441 +0x84

Previous read at 0x00c0019a0a88 by goroutine 2438:
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*spanContext).setSamplingPriority()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/spancontext.go:185 +0x1bc
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*span).setSamplingPriorityLocked()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/span.go:263 +0x8c
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*span).setTagBool()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/span.go:388 +0xdc
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*span).SetTag()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/span.go:127 +0x134
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.TestTraceManualKeepRace.func1.1()
      /Users/evan.jones/dd/dd-trace-go/ddtrace/tracer/span_test.go:442 +0xa8
```

### Motivation

Fix flaky tests in programs that use dd-trace-go.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: ddtrace/tracer/spancontext.go [hunk 1]
```diff
@@ -290,10 +290,11 @@ func (t *trace) samplingPriority() (p int, ok bool) {
 	return t.samplingPriorityLocked()
 }
 
-func (t *trace) setSamplingPriority(p int, sampler samplernames.SamplerName) {
+// setSamplingPriority sets the sampling priority and returns true if it was modified.
+func (t *trace) setSamplingPriority(p int, sampler samplernames.SamplerName) bool {
 	t.mu.Lock()
 	defer t.mu.Unlock()
```
Comment (darccio) [inline ddtrace/tracer/spancontext.go:296]: Reviewing this it seems to me that the original issue comes from checking `c.trace.priority != nil && *c.trace.priority != float64(p)` outside of the highlighted mutex.

In general, the change seems safe.

**Review by darccio — APPROVED** [review]

File: ddtrace/tracer/span_test.go [hunk 2]
```diff
@@ -387,6 +388,53 @@ func TestTraceManualKeepAndManualDrop(t *testing.T) {
...
+		wg.Add(numGoroutines)
+		for j := 0; j < numGoroutines; j++ {
+			go func() {
+				defer wg.Done()
+				childSpan := tracer.StartSpan(
+					"child",
+					ChildOf(rootSpan.Context()),
+					Tag(ext.ManualKeep, true),
+				)
+				defer childSpan.Finish()
```
[the remark this answers is not in the document]
Comment (evanj) [reply ddtrace/tracer/span_test.go:431]: I think there is a previous comment missing, or at least I can't seem to see it, but I can infer that it was to remove the defers. Good idea! I've just submitted that change.

File: ddtrace/tracer/spancontext.go (hunk 1 above)
Comment (evanj) [reply ddtrace/tracer/spancontext.go:296]: That is correct: this fix ensures we only read `t.priority` while holding the lock, which required returning the boolean from this function.

**Review by evanj — COMMENTED** [review]
Thanks for the review. I have applied your suggestions!

**Comment by evanj** [comment]
This was flagged stale but I believe this probably can be merged? @darccio or others, is there anything else I can do to help here?

**Comment by darccio** [comment]
@evanj I think we can merge. Let me update the branch, run the tests and check everything goes fine.

**Pull request merged**